osu.Game.Rulesets.Osu/Mods/OsuModRelax.cs

```diff
@@ -101,7 +104,7 @@ public void Update(Playfield playfield)
 
             if (requiresHold)
                 changeState(true);
-            else if (isDownState && time - lastStateChangeTime > AutoGenerator.KEY_UP_DELAY)
+            else if ((isDownState && time + AutoGenerator.KEY_UP_DELAY >= beatmapLength) || time - lastStateChangeTime > AutoGenerator.KEY_UP_DELAY)
                 changeState(false);
```

Comment:
Along with the above, this conditional is very hard to read/parse. Based on the issue you stated to fix, isn't this what you want to do (with some extracting to potentially improve readability):
```suggestion
            else
            {
                 bool hasCompleted = time >= beatmapLength;
                 bool shouldRelease = isDownState && (time - lastStateChangeTime > AutoGenerator.KEY_UP_DELAY || hasCompleted);
                 
                 if (shouldRelease)
                     changeState(false);
            }
```